### PR TITLE
Testrtc linechart

### PR DESCRIPTION
**Pull request by andresusanopinto**
This is quite experimental.. but I think it would beneficial to submit and improve over it instead of being rotting in my pull requests.

For now this plot is only used in WiFi periodic test which is an explicit test so it should not ruin any other user experience.

### CODE REVIEW

File: samples/web/content/testrtc/ui/line-chart.html [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        this.x.push(timestamp / 100.0 * 2);
+        this.y.push(this.calculateY(value));
+
+        if (this.xStart === null || this.x[index] - this.xStart > this.$.canvas.width) {
+          // First point added or a point added that clears the chart.
+          this.setupChart();
+          this.xStart = this.x[index] - this.xOffset;
+        } else {
+          this.ctx.strokeStyle = (Math.min(this.y[index], this.y[index-1]) < 50 ? 'red': 'green');
+          this.ctx.beginPath();
```
Comment (KaptenJansson) [inline samples/web/content/testrtc/ui/line-chart.html]: Could you make the [peak](https://drive.google.com/a/google.com/file/d/0BwwlrcPnEG-RSzBidHZSSGZ4dzg/view?usp=sharing) be color matched on both sides?

File: samples/web/content/testrtc/ui/line-chart.html [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        }
+      },
+
+      setupChart: function () {
+        this.ctx.clearRect (0 , 0, this.$.canvas.width, this.$.canvas.height);
+        this.ctx.lineWidth = window.devicePixelRatio;
+        this.ctx.save();
+
+        this.ctx.lineWidth = window.devicePixelRatio;
+        this.ctx.strokeStyle = 'lightgray';
```
Comment (KaptenJansson) [inline samples/web/content/testrtc/ui/line-chart.html]: For styles it should be 3 hex color codes where possible but since this is an html file it's fine.

**Comment by KaptenJansson** [comment]
Consider renaming the test to something like "Network latency analyzer"

File: samples/web/content/testrtc/ui/line-chart.html [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    var interval = 0;
+    Polymer({
+      x: [],
+      y: [],
+      yScale: [0, 600],
+      yLabels: [0, 50, 100, 150, 200, 250, 300, 350, 400, 450, 500],
+      xStart: null,
+      xOffset: null,
+      lastSlowRender: null,
+      addDatapoint: function (timestamp, value) {
```
Comment (KaptenJansson) [inline samples/web/content/testrtc/ui/line-chart.html]: Remove space after function.

File: samples/web/content/testrtc/ui/line-chart.html [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        var start = window.performance.now();
+        this.ctx = this.$.canvas.getContext('2d');
+
+        var index = this.x.length;
+        this.x.push(timestamp / 100.0 * 2);
+        this.y.push(this.calculateY(value));
+
+        if (this.xStart === null || this.x[index] - this.xStart > this.$.canvas.width) {
```
Comment (KaptenJansson) [inline samples/web/content/testrtc/ui/line-chart.html]: Consider linewrap this as it's longer than 80 chars.

File: samples/web/content/testrtc/ui/line-chart.html [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+          // First point added or a point added that clears the chart.
+          this.setupChart();
+          this.xStart = this.x[index] - this.xOffset;
+        } else {
+          this.ctx.strokeStyle = (Math.min(this.y[index], this.y[index-1]) < 50 ? 'red': 'green');
```
Comment (KaptenJansson) [inline samples/web/content/testrtc/ui/line-chart.html]: Consider linewrap this as it's longer than 80 chars.

File: samples/web/content/testrtc/ui/line-chart.html [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+          this.ctx.beginPath();
+          this.ctx.moveTo(this.x[index - 1] - this.xStart, this.y[index - 1]);
+          this.ctx.lineTo(this.x[index] - this.xStart, this.y[index]);
+          this.ctx.stroke();
+          this.ctx.closePath();
+        }
+
+        var now = window.performance.now();
+        var interval = now - start;
+        if (interval > 5 && (this.lastSlowRender === null || (now - this.lastSlowRender) > 5000)) {
```
Comment (KaptenJansson) [inline samples/web/content/testrtc/ui/line-chart.html]: Consider linewrap this as it's longer than 80 chars.

File: samples/web/content/testrtc/ui/line-chart.html [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+          this.lastSlowRender = now;
+          console.log('Updating the graph was slow and might affect results. Graph update took ' + interval + ' ms');
```
Comment (KaptenJansson) [inline samples/web/content/testrtc/ui/line-chart.html]: Consider linewrap this as it's longer than 80 chars.

File: samples/web/content/testrtc/ui/line-chart.html [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+          this.ctx.translate(0.5, 0.5);
+        }
+
+        this.ctx.textAlign = 'right';
+        this.ctx.textBaseline = 'middle';
+
+        this.ctx.beginPath();
+        this.xOffset = 0;
+        for (var index = 0; index != this.yLabels.length; index++) {
+          this.xOffset = Math.max(this.xOffset, this.ctx.measureText(this.yLabels[index] + ' ms ').width);
```
Comment (KaptenJansson) [inline samples/web/content/testrtc/ui/line-chart.html]: Consider linewrap this as it's longer than 80 chars.

File: samples/web/content/testrtc/ui/line-chart.html [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+ *  Use of this source code is governed by a BSD-style license
+ *  that can be found in the LICENSE file in the root of the source
+ *  tree.
+-->
+
+<!--
+`<line-chart>` is used to display a line chart updating in real-time to the user
+in an efficient way and without locking the js thread for too long.
+
+It is not a very advanced or a pretty line chart render, but it is very specialized
```
Comment (KaptenJansson) [inline samples/web/content/testrtc/ui/line-chart.html]: Consider linewrap this as it's longer than 80 chars.

**Comment by KaptenJansson** [comment]
Also consider adding a timeout with an error for the test if the network connection is disconnected suddenly.

File: samples/web/content/testrtc/ui/line-chart.html (hunk 4 above)
Comment (andresusanopinto) [inline samples/web/content/testrtc/ui/line-chart.html]: Done.

File: samples/web/content/testrtc/ui/line-chart.html (hunk 5 above)
Comment (andresusanopinto) [inline samples/web/content/testrtc/ui/line-chart.html]: Done.

File: samples/web/content/testrtc/ui/line-chart.html (hunk 6 above)
Comment (andresusanopinto) [inline samples/web/content/testrtc/ui/line-chart.html]: Done.

File: samples/web/content/testrtc/ui/line-chart.html (hunk 7 above)
Comment (andresusanopinto) [inline samples/web/content/testrtc/ui/line-chart.html]: Done.

File: samples/web/content/testrtc/ui/line-chart.html (hunk 8 above)
Comment (andresusanopinto) [inline samples/web/content/testrtc/ui/line-chart.html]: Done.

File: samples/web/content/testrtc/ui/line-chart.html (hunk 9 above)
Comment (andresusanopinto) [inline samples/web/content/testrtc/ui/line-chart.html]: Done.

File: samples/web/content/testrtc/ui/line-chart.html (hunk 3 above)
Comment (andresusanopinto) [inline samples/web/content/testrtc/ui/line-chart.html]: Done.

File: samples/web/content/testrtc/ui/line-chart.html (hunk 1 above)
Comment (andresusanopinto) [inline samples/web/content/testrtc/ui/line-chart.html]: That is a bit hard... instead I changed the line to be grey and dots on the samples to be either green or red depending on whether they are < 300ms.

File: samples/web/content/testrtc/ui/line-chart.html (hunk 2 above)
Comment (andresusanopinto) [inline samples/web/content/testrtc/ui/line-chart.html]: I would like to leave it named colors as it is easier to red.

**Comment by andresusanopinto** [comment]
Renaming and timeout on test for users who get disconnected is left for another CL.

**Comment by KaptenJansson** [comment]
LGTM

**Pull request merged**